Add isInteger and toInteger operations to ContourNumber

ronsilverentand opened the pull request:
Add isInteger and toInteger operations to ContourNumber for validation and transformation

number.dart
isInteger and to Integer added

number_test.dart
4 additional tests
- isInteger should add a isInteger operation
- isInteger should validate number value
- toInteger should add a toInteger operation
- toInteger should validate number value

Fixes #58 and #59 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Enhanced initialization: Users can now supply an optional initial value when creating instances for common data types (booleans, lists, numbers, objects, and strings).
  - Improved numeric support: New operations allow for integer validation and conversion to ensure numeric values are processed accurately.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ronsilverentand committed:
Add isInteger and toInteger operations to ContourNumber for validation and transformation

ronsilverentand committed:
Rename test cases to reflect the toInteger operation for clarity

ronsilverentand committed:
Add optional value parameter to instance methods and update tests for initial value handling

ronsilverentand commented:
Also added solution for issue #59 
initial value functionality for instance added to Boolean, Number, String, List and Object

lucasilverentand commented on packages/contour/lib/src/types/number.dart:
We should maybe merge it into one: 
```
.integer(force: bool)
```

when force is true it is a transform
when force is false it is a check

ronsilverentand replied:
Sounds valid. Will adapt tomorrow

lucasilverentand commented on packages/contour/test/src/types/object_test.dart:
```diff
@@ -34,13 +34,17 @@ void main() {
       final objectType = ContourObject({
         'name': ContourString(),
         'age': ContourNumber(),
       }).additionalFields(false);
```
Perhaps we should give .additionalFields parameters as well just like I'm suggesting for integer, where it would have .additionalFields(force: bool) 

force = true, means strip them out and do not throw an error
force = false, means do a check and throw an error if fields are found

What do you think @ronsilverentand?

ronsilverentand replied:
Sounds valid. Will resolve tomorrow

lucasilverentand reviewed (changes requested):
In general everything is correct and looks good but perhaps a couple parameters could be added, looking for your feedback @ronsilverentand

lucasilverentand commented:
@coderabbit review

lucasilverentand commented on packages/contour/test/src/types/boolean_test.dart:
```diff
@@ -88,4 +88,16 @@ void main() {
     });
   });
+
+  test('initial value should be null', () {
+    final numberType = ContourBoolean();
+    final instance = numberType.instance('test');
+    expect(instance.value, isNull);
+  });
+
+  test('initial value should be set', () {
+    final numberType = ContourBoolean();
+    final instance = numberType.instance('test', true);
+    expect(instance.value, true);
+  });
```
not needed for now as test amount is low enough

lucasilverentand committed:
Apply suggestions from code review

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Luca Silverentand <luca@onezero.company>

lucasilverentand committed:
Update packages/contour/lib/src/types/number.dart

Improve null handling in isInteger check as previous solution would handle null incorrectly

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>
Signed-off-by: Luca Silverentand <luca@onezero.company>

ronsilverentand committed:
Refactor ContourNumber operations to include integer validation and transformation; update tests for additionalFields validation in ContourObject

ronsilverentand committed:
Merge branch '58-add-integer-checktransform-on-number-in-contour' of https://github.com/onezerocompany/oui into 58-add-integer-checktransform-on-number-in-contour
